### PR TITLE
ARROW-5463: [Rust] Add AsRef trait for Buffer.

### PR DESCRIPTION
**Pull request by liurenjie1024**
This is useful when using Buffer as typed slices. @sunchao @nevi-me @andygrove Please help to review when you get a chance.

### CODE REVIEW

**Review by nevi-me — APPROVED** [review]
Hi @liurenjie1024, I played around with `make_as_ref!()` macro to see if I could convert it to a generic function, and couldn't, so I'm happy with the macro unless someone else can make a generic fn work.

**Comment by liurenjie1024** [comment]
@nevi-me  I also tried and couldn't figure out a fancy way to it, so just used macro.

**Comment by sunchao** [comment]
I think the problem is the `u8` case which will cause a conflict. As an alternative, we can instead implement a function similar to the existing `data`, but returns an slice of `T` instead of `u8`:
```rust
pub fn typed_data<T: ArrowNativeType>(&self) -> &[T] { ... }
```

This will save from the macros and in terms of usage it should be similar.

**Comment by liurenjie1024** [comment]
@sunchao There are two conflict cases.

1. AsRef<[u8]> will add another From implementation for Buffer, since we already have `impl<T: AsRef<[u8]>> From<T> for Buffer` implementation.
2. I don't how to implement AsRef<[bool]> for Buffer since the implementation logic is not trivial.

**Comment by sunchao** [comment]
@liurenjie1024 In that case maybe you can change the above code to:
```rust
pub fn typed_data<T: ArrowNativeType + num::Num>(&self) -> &[T] { ... }
```

The problem against macros is that whenever someone adds a new type he/she needs to know all the places to modify to cover the places that the type is used.

**Comment by liurenjie1024** [comment]
@sunchao It helps with bool type. But if we need to add support for u8, we need to remove this piece of code:
```rust
/// Creating a `Buffer` instance by copying the memory from a `AsRef<[u8]>` into a newly
/// allocated memory region.
impl<T: AsRef<[u8]>> From<T> for Buffer {
    fn from(p: T) -> Self {
        // allocate aligned memory buffer
        let slice = p.as_ref();
        let len = slice.len() * mem::size_of::<u8>();
        let capacity = bit_util::round_upto_multiple_of_64(len);
        let buffer = memory::allocate_aligned(capacity);
        unsafe {
            memory::memcpy(buffer, slice.as_ptr(), len);
        }
        Buffer::from_raw_parts(buffer, len)
    }
}
```

**Comment by sunchao** [comment]
Hmm.. why is that? this only happens if you implement `AsRef[u8]` for `Buffer` right? if you use the `typed_data` approach you should not need it.

**Comment by liurenjie1024** [comment]
Yes, you are right. Changed function name to typed _data.

File: rust/arrow/src/buffer.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@
 use packed_simd::u8x64;
 
 use std::cmp;
+use std::convert::AsRef;
```
Comment (sunchao) [inline rust/arrow/src/buffer.rs]: nit: this is not used?

**Review by sunchao — APPROVED** [review]

File: rust/arrow/src/buffer.rs (hunk 1 above)
Comment (liurenjie1024) [reply rust/arrow/src/buffer.rs]: This seems to be used by other parts. Anyway, cargo check doesn't complain about this.

Comment (sunchao) [reply rust/arrow/src/buffer.rs]: I see. I think it's in the preclude so no need to explicitly import previously.

**Pull request closed without merging**